modules/react/common/spec/components.test-d.tsx

```diff
@@ -0,0 +1,51 @@
+import React from 'react';
+import {expectTypeOf} from 'expect-type';
+
+import {createComponent, ElementComponent} from '../lib/utils/components';
+
+it('should assign an element-base component as an ElementComponent', () => {
+  const component = createComponent('div')({Component: (props: {foo: 'bar'}) => null});
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}>>();
+});
+
+it('should add sub-components to the signature', () => {
+  const component = createComponent('div')({
+    Component: (props: {foo: 'bar'}) => null,
+    subComponents: {
+      Foo: 'bar',
+    },
+  });
+  expectTypeOf(component).toEqualTypeOf<ElementComponent<'div', {foo: 'bar'}> & {Foo: number}>();
+});
+
+it('should assign ref and Element correctly for element components', () => {
+  createComponent('div')({
+    Component: (props: {}, ref, Element) => {
+      expectTypeOf(ref).toEqualTypeOf<React.Ref<HTMLDivElement>>();
+      expectTypeOf(Element).toEqualTypeOf<'div'>();
+      return null;
+    },
+  });
+});
+
+it('should assign ref and Element correctly for createComponent components', () => {
+  const component = createComponent('article')({Component: (props: {}) => null});
+
+  createComponent(component)({
+    Component: (props: {}, ref, Element) => {
+      expectTypeOf(ref).toEqualTypeOf<React.Ref<HTMLElement>>();
+      expectTypeOf(Element).toEqualTypeOf<ElementComponent<'article', {}>>();
```

Comment:
This is kind of a cool way to verify what gets returned by functions in an explicit way.